File: GalleryContext.BusinessLogic/Models/ValueObjects/ArtworkName.cs
using System.Collections.Generic;
using GalleryContext.BusinessLogic.Errors;
using SharedKernel.Core.Primitives;

namespace GalleryContext.BusinessLogic.Models.ValueObjects;

public sealed class ArtworkName : ValueObject
{
    public const int MinLength = 3;
    public const int MaxLength = 255;
    
    public string Value { get; }

    private ArtworkName(string value)
    {
        Value = value;
    }

    public static Result<ArtworkName> Create(string name)
    {
        if (string.IsNullOrWhiteSpace(name))
        {
            return Result<ArtworkName>.Failure(DomainErrors.Artwork.NameRequired);
        }

        if (name.Length < MinLength)
        {
            return Result<ArtworkName>.Failure(DomainErrors.Artwork.NameTooShort);
        }

        if (name.Length > MaxLength)
        {
            return Result<ArtworkName>.Failure(DomainErrors.Artwork.NameTooLong);
        }

        return Result<ArtworkName>.Success(new ArtworkName(name));
    }
    internal static ArtworkName Hydrate(string name)
    {
        return new ArtworkName(name);
    }
    public override IEnumerable<object> GetAtomicValues()
    {
        yield return Value;
    }
    
    public static implicit operator string(ArtworkName name) => name.Value;
}

File: Client/lajemacarts/src/app/features/dashboard/domain/models/value-objects/artwork-name.ts
import {Result} from '../../../../../shared/core/result';
import {DomainErrors} from '../../../../../shared/domain/errors/domain-errors';

export class ArtworkName {
  public static readonly MinLength = 3;
  public static readonly MaxLength = 255;

  public readonly value: string;

  private constructor(value: string) {
    this.value = value;
  }

  public static create(name: string): Result<ArtworkName> {
    if (!name?.trim()) {
      return Result.failure<ArtworkName>(DomainErrors.Artwork.NameRequired);
    }
    if (name.length < this.MinLength) {
      return Result.failure<ArtworkName>(DomainErrors.Artwork.NameTooShort);
    }
    if (name.length > this.MaxLength) {
      return Result.failure<ArtworkName>(DomainErrors.Artwork.NameTooLong);
    }
    return Result.success<ArtworkName>(new ArtworkName(name));
  }


  public static hydrate(value: string): ArtworkName {
    return new ArtworkName(value);
  }
}
